lib/internal/streams/add-abort-signal.js

```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+const {
+  AbortError,
+  codes,
+} = require('internal/errors');
+
+const { ERR_INVALID_ARG_TYPE } = codes;
+
+const {
+  validateAbortSignal,
+} = require('internal/validators');
+
+function isReadable(obj) {
+  return !!(obj && typeof obj.pipe === 'function');
+}
+
+module.exports = function addAbortSignal(signal, readable) {
+  validateAbortSignal(signal, 'signal');
+  if (!isReadable(readable)) {
+    throw new ERR_INVALID_ARG_TYPE('readable', 'stream.Readable', readable);
+  }
+  const onAbort = () => {
+    readable.destroy(new AbortError());
+  };
+  if (signal.aborted) {
+    onAbort();
+  } else {
+    signal.addEventListener('abort', onAbort);
+    readable.on('close', () => signal.removeEventListener('abort', onAbort));
```

Comment:
This will not work for "legacy" streams.

Reply:
Yea, probably a good idea to use `stream.finished` here.